### PR TITLE
Add chain RPC calls

### PR DESCRIPTION
**Pull request by cubismic**
I implemented some RPC calls.
https://github.com/CodeChain-io/codechain-sdk-js/issues/283

There are some RPC calls that are not implemented yet.

### CODE REVIEW

File: src/rpc/chain.ts [hunk 1]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
                 .catch(reject);
         });
     }
+
+    /**
+     * Gets the number of shards, at the state of given blockNumber
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: Gets the number of shards, at the state of the given blockNumber

File: src/rpc/chain.ts [hunk 2]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+                            )
+                        );
+                    }
+                })
+                .catch(reject);
+        });
+    }
+
+    /**
+     * Gets the root of shard, at the state of the given blockNumber.
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: Gets the root of the shard

File: src/rpc/chain.ts [hunk 3]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+     * @param shardId A shard Id.
+     * @param blockNumber A block number.
+     * @returns The hash of root of shard.
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: The hash of the root of the shard

File: src/rpc/chain.ts [hunk 4]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+        blockNumber?: number
+    ): Promise<H256 | null> {
+        if (!isShardIdValue(shardId)) {
+            throw Error(
+                `Expected the second argument of getAssetSchemeByTracker to be a shard ID value but found ${shardId}`
+            );
+        }
+        if (blockNumber !== undefined && !isNonNegativeInterger(blockNumber)) {
+            throw Error(
+                `Expected the third argument of getAssetSchemeByTracker to be non-negative integer but found ${blockNumber}`
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: Expected the third argument of getAssetSchemeByTracker to be a non-negative integer but found

File: src/rpc/chain.ts [hunk 5]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+
+    /**
+     * Gets the mining reward of the given block number.
+     * @param blockNumber A block nubmer.
+     * @returns The amount of mining reward, or null if the given block number is not mined yet.
+     */
+    public getMiningReward(blockNumber: number): Promise<U64 | null> {
+        if (!isNonNegativeInterger(blockNumber)) {
+            throw Error(
+                `Expected the third argument of getAssetSchemeByTracker to be non-negative integer but found ${blockNumber}`
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: Expected the third argument of getAssetSchemeByTracker to be a non-negative integer but found

File: src/rpc/chain.ts [hunk 6]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+        return new Promise((resolve, reject) => {
+            this.rpc
+                .sendRpcRequest("chain_executeTransaction", [tx, sender])
+                .then(result => {
+                    if (typeof result === "boolean") {
+                        resolve(result);
+                    } else {
+                        reject(
+                            Error(
+                                `Expected chain_isAssetSpent to return either null or boolean but it returned ${result}`
```
Comment (ScarletBlue) [inline src/rpc/chain.ts]: Expected chain_isAssetSpent to return either null or a boolean but it returned

File: src/rpc/chain.ts [hunk 7]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+        return new Promise((resolve, reject) => {
+            this.rpc
+                .sendRpcRequest("chain_executeTransaction", [tx, sender])
+                .then(result => {
+                    if (typeof result === "boolean") {
+                        resolve(result);
+                    } else {
+                        reject(
+                            Error(
+                                `Expected chain_isAssetSpent to return either null or a boolean but it returned ${result}`
```
Comment (joojis) [inline src/rpc/chain.ts]: - return a boolean
- chain_exeucteTransaction

File: src/rpc/chain.ts [hunk 8]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+
+    /**
+     * Executes the transactions.
+     * @param tx A transaction to execute.
+     * @param sender A platform address of sender.
+     * @returns True, if the transaction is executed successfully. False, if the transaction is not executed.
+     */
+    public executeTransaction(
+        tx: Transaction,
+        sender: PlatformAddress
```
Comment (joojis) [inline src/rpc/chain.ts]: PlatformAddress | string

File: src/rpc/chain.ts [hunk 9]
```diff
@@ -999,6 +999,170 @@ export class ChainRpc {
...
+            );
+        }
+        if (!PlatformAddress.check(sender)) {
+            throw Error(
+                `Expected the second argument of executeTransaction to be a PlatformAddress value but found ${PlatformAddress}`
+            );
+        }
+        return new Promise((resolve, reject) => {
+            this.rpc
+                .sendRpcRequest("chain_executeTransaction", [tx, sender])
```
Comment (joojis) [inline src/rpc/chain.ts]: Have you tested this code? It doesn't seem to work.

File: examples/execute-tx.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        shardId,
+        metadata: JSON.stringify({
+            name: "Gold",
+            description: "An asset example",
+            icon_url: "https://gold.image/"
+        }),
+        supply: 10000,
+        administrator: ACCOUNT_ADDRESS
+    });
+    const alpha = sdk.core.createMintAssetTransaction({
```
Comment (joojis) [inline examples/execute-tx.js]: tx

File: src/rpc/chain.ts [hunk 11]
```diff
@@ -999,6 +999,173 @@ export class ChainRpc {
...
+        return new Promise((resolve, reject) => {
+            this.rpc
+                .sendRpcRequest("chain_getNumberOfShards", [blockNumber])
+                .then(result => {
+                    if (result === null || typeof result === "number") {
+                        resolve(result);
+                    } else {
+                        reject(
+                            Error(
+                                `Expected chain_getNumberOfShards to return either null or number, but it returned ${result}`
```
Comment (joojis) [inline src/rpc/chain.ts]: `return either null or number` -> `return a number`

File: src/rpc/chain.ts [hunk 12]
```diff
@@ -999,6 +999,173 @@ export class ChainRpc {
...
+     * @param blockNumber A block number.
+     * @returns The hash of root of the shard.
+     */
+    public getShardRoot(
+        shardId: number,
+        blockNumber?: number
+    ): Promise<H256 | null> {
+        if (!isShardIdValue(shardId)) {
+            throw Error(
+                `Expected the first argument of getAssetSchemeByTracker to be a shard ID value but found ${shardId}`
```
Comment (joojis) [inline src/rpc/chain.ts]: `getAssetSchemeByTracker` -> `getShardRoot`

File: src/rpc/chain.ts [hunk 13]
```diff
@@ -999,6 +999,173 @@ export class ChainRpc {
...
+            );
+        }
+        if (blockNumber !== undefined && !isNonNegativeInterger(blockNumber)) {
+            throw Error(
+                `Expected the second argument of getAssetSchemeByTracker to be a non-negative integer but found ${blockNumber}`
```
Comment (joojis) [inline src/rpc/chain.ts]: `getAssetSchemeByTracker` -> `getShardRoot`

File: src/rpc/chain.ts [hunk 14]
```diff
@@ -999,6 +999,173 @@ export class ChainRpc {
...
+        return new Promise((resolve, reject) => {
+            this.rpc
+                .sendRpcRequest("chain_getMiningReward", [blockNumber])
+                .then(result => {
+                    try {
+                        resolve(result === null ? null : new U64(result));
+                    } catch (e) {
+                        reject(
+                            Error(
+                                `Expected chain_getShardRoot to return either null or a value of U64, but an error occurred: ${e.toString()}`
```
Comment (joojis) [inline src/rpc/chain.ts]: `chain_getShardRoot` -> `chain_getMiningReward`

**Pull request merged**